kivy/app.py

```diff
@@ -737,6 +740,29 @@ def directory(self):
                 self._app_directory = '.'
         return self._app_directory
 
+    def _get_user_data_dir(self):
+        # Determine and return the user_data_dir.
+        data_dir = ""
+        if platform == 'ios':
+            data_dir = join('~/Documents', self.name)
+        elif platform == 'android':
+            from jnius import autoclass, cast
+            PythonActivity = autoclass('org.kivy.android.PythonActivity')
+            context = cast('android.content.Context', PythonActivity.mActivity)
+            file_p = cast('java.io.File', context.getFilesDir())
+            data_dir = file_p.getAbsolutePath()
+        elif platform == 'win':
+            data_dir = os.path.join(os.environ['APPDATA'], self.name)
+        elif platform == 'macosx':
+            data_dir = '~/Library/Application Support/{}'.format(self.name)
+        else:  # _platform == 'linux' or anything else...:
+            data_dir = os.environ.get('XDG_CONFIG_HOME', '~/.config')
+            data_dir = join(data_dir, self.name)
+        data_dir = expanduser(data_dir)
```

Comment:
the location of this line seems like a bug, not caused by this PR of course but maybe worth fixing in the same go? `expanduser` is not correct to do on the Windows and Android paths, and might actually lead to this function returning an entirely wrong path on Android **since it is possible to create a folder with `~` in the name on Linux** (and therefore most likely on Android)!

So this line should really be moved inside the cases where it is appropriate, which as far as I can tell is Linux, iOS, macOS and possibly Linux - although here it might also be wrong, depending on whether is `XDG_CONFIG_HOME` allowed to contain `~` which I can't tell off the cuff. Might be worth checking the freedesktop spec

Reply:
Glad you mentioned that. I noticed that too but decided not to play with it. But I agree, now is probably a good time to change that. Will do...

Reply:
Nice :+1: I checked the new `expanduser` locations and from my understanding and knowledge, they are correct now. Very good! :tada: